llvm/lib/Target/MSP430/MSP430EpilogueOptimization.cpp

```diff
@@ -0,0 +1,142 @@
+///
+/// This pass is expected to be run after epilogue generation is done and
+/// is only applicable to MSP430 (not MSP430X).
+///
+/// See MSP430 Embedded Application Binary Interface, part 3.8
+/// "__mspabi_func_epilog Helper Functions" for details.
+//===----------------------------------------------------------------------===//
+
+#include "MSP430.h"
+#include "MSP430TargetMachine.h"
+
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+
+using namespace llvm;
+
+namespace {
+
+class MSP430EpilogueOptimization : public MachineFunctionPass {
+public:
+  MSP430EpilogueOptimization(): MachineFunctionPass(ID) {}
+
+  bool runOnMachineFunction(MachineFunction &MF) override;
+private:
+  static char ID;
+  const MSP430InstrInfo *TII;
+
+  void replaceTailWithJmpToLibCall(MachineBasicBlock &MBB,
+                                   MachineInstr &FirstInstrToReplace,
+                                   const char *LibCallName);
+
+  bool tryReplacingEpilogue(MachineBasicBlock &MBB);
+};
+
+char MSP430EpilogueOptimization::ID = 0;
+
+} // end anonymous namespace
+
+/// \brief LibCall names arranged so that `N`-th element restores N registers
+///
+/// MSP430 EABI defines standard epilogue helper functions like these:
+/// \code
+/// __mspabi_func_epilog_7: RET R4
+/// __mspabi_func_epilog_6: RET R5
+/// __mspabi_func_epilog_5: RET R6
+/// __mspabi_func_epilog_4: RET R7
+/// __mspabi_func_epilog_3: RET R8
+/// __mspabi_func_epilog_2: RET R9
+/// __mspabi_func_epilog_1: RET R10
+///                         RET
+/// \endcode
+/// @note These helper functions do not exist on MSP430X
+static const char *EpilogueLibCallNames[] = {
+  nullptr,
+  "__mspabi_func_epilog_1",
+  "__mspabi_func_epilog_2",
+  "__mspabi_func_epilog_3",
+  "__mspabi_func_epilog_4",
+  "__mspabi_func_epilog_5",
+  "__mspabi_func_epilog_6",
+  "__mspabi_func_epilog_7",
+};
+
+/// Registers to be restored, from the last `POP` instruction to the first
+static const unsigned RestoredRegisters[] = {
+  MSP430::R10, MSP430::R9, MSP430::R8, MSP430::R7,
+  MSP430::R6, MSP430::R5, MSP430::FP /* R4 */,
+};
+
+void MSP430EpilogueOptimization::replaceTailWithJmpToLibCall(
+    MachineBasicBlock &MBB, MachineInstr &FirstInstrToReplace,
+    const char *LibCallName) {
+
+  const DebugLoc &DL = FirstInstrToReplace.getDebugLoc();
+  BuildMI(MBB, FirstInstrToReplace, DL, TII->get(MSP430::JMP))
+      .addExternalSymbol(LibCallName);
+
+  for (MachineInstr *MI = &FirstInstrToReplace; MI != nullptr; ) {
+    MachineInstr *InstrToRemove = MI;
+    MI = MI->getNextNode();
+    InstrToRemove->eraseFromParent();
+  }
+}
+
+bool MSP430EpilogueOptimization::tryReplacingEpilogue(MachineBasicBlock &MBB) {
+  MachineInstr *LastMatchedPopInstr = nullptr;
+  unsigned MatchedRegisters = 0;
+
+  MachineBasicBlock::iterator Terminator = MBB.getFirstTerminator();
+  if (Terminator == MBB.end() || Terminator->getOpcode() != MSP430::RET) {
```

Comment:
Typically LLVM style guide suggests not to use braces on 1 line bodies unless it would be ambiguous. Same below